### ISSUE TITLE
Joining swarm a second time as a worker after promotion made the node a manager

### ISSUE THREAD

**Issue by aaronlehmann** [docker/docker#28490]
I tried to add a second manager using a dind setup. I added the new node as a worker and then promoted it. I forgot to set the listen address to something that would not conflict. The manager did not start successfully, and I tried to remove the node from the swarm and try again. When I rejoined the cluster, the node was a manager even though I used the worker join token. It seems to have retained the same certificate as before.

```
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker -H unix:///var/run/docker2.sock swarm join --token <worker token> 172.17.0.3:2377
This node joined a swarm as a worker.
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker node ls
ID                           HOSTNAME      STATUS  AVAILABILITY  MANAGER STATUS
6y2eiflcajy8ncnjfoi0joew4    56cdb93b9fa5  Ready   Active
dzd9a6s7dsggoe5nih23bcdtg *  56cdb93b9fa5  Ready   Active        Leader
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker node promote 6y2eiflcajy8ncnjfoi0joew4
Node 6y2eiflcajy8ncnjfoi0joew4 promoted to a manager in the swarm.
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker -H unix:///var/run/docker2.sock swarm leave
Error response from daemon: listen tcp 0.0.0.0:2377: bind: address already in use
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker -H unix:///var/run/docker2.sock swarm leave --force
Error response from daemon: This node is not part of a swarm
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker node ls
ID                           HOSTNAME      STATUS  AVAILABILITY  MANAGER STATUS
6y2eiflcajy8ncnjfoi0joew4    56cdb93b9fa5  Down    Active
dzd9a6s7dsggoe5nih23bcdtg *  56cdb93b9fa5  Ready   Active        Leader
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker -H unix:///var/run/docker2.sock swarm join --token <worker token>  --listen-addr 172.17.0.3:2378   172.17.0.3:2377
This node joined a swarm as a manager.
root@56cdb93b9fa5:/go/src/github.com/docker/docker# docker node ls
ID                           HOSTNAME      STATUS  AVAILABILITY  MANAGER STATUS
6y2eiflcajy8ncnjfoi0joew4    56cdb93b9fa5  Ready   Active        Reachable
dzd9a6s7dsggoe5nih23bcdtg *  56cdb93b9fa5  Ready   Active        Leader
```

```
# docker version
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 21:39:14 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 21:39:14 2016
 OS/Arch:      linux/amd64
```

cc @tonistiigi @LK4D4 @diogomonica

**Comment by aluzzardi** [comment]
ping @diogomonica 

I don't know if this is a P0, but a worker became manager. Please downgrade to P2 if you don't think it's a big issue

**Comment by diogomonica** [comment]
This is less concerning since the node had been a manager first, and arguably all of your data has been leaked already if this node has had access to it and was misbehaving for you to demote it, but worth keeping as a P0 until we figure out what it is.

**Comment by tonistiigi** [comment]
This happened because the node never left swarm, `swarm leave` never returned 0 exit code. The bug is that 1) leave of a node in error state should have worked and removed the state 2) join of a node in error state should have not succeeded(atm it just updated node's advertise address and caused it to start properly).

These are fixed in #28511 but I'll make a better refactor to properly handle the node in error state.

No sign of any security problems like allowing nodes to become manager without proper certificate.

**Comment by diogomonica** [comment]
@tonistiigi thanks for taking a look. Are you still going to go through with a better refactor?

**Comment by tonistiigi** [comment]
@diogomonica Yes, but it looks to be quite extensive.

**Comment by diogomonica** [comment]
@tonistiigi given the current state of the code, I'm fine with that. Hopefully it's more readable, and makes less use of routines an channels.

**Comment by tonistiigi** [comment]
This should be fixed by #28907
